### ISSUE TITLE
Windows and non-Latin text

### ISSUE THREAD

**Issue by the-Arioch** [castle-engine/castle-engine#360]
![2022-01-09 15_32_45-Р—РІСѓРє РІ РІРѕРґРµ](https://user-images.githubusercontent.com/937774/148682897-2fccf77a-d8ca-4029-8a4b-e192ff4a0385.png)

Look at the title bar...

`CastleEngineManifest.xml` contains correct text in Russian.
But then the Castle sets main window Caption to that byte stream as is, without converting.

Win32/Win64 API for every "semantic" function usually have two ABI functions, FunctionNameA and FunctionNameW.
Latter takes texts in URF-16 encoding (UCS2 in old WinNT versions).
Former takes "MBCS" texts which usually correspond with "ANSI codepage" for the given language ("user locale") providing we talk about GUI application.

_For text-mode applications i am not so sure, their UI would use "OEM codepage" instead, btut funcitons like file i/o? dunno, it is easy to test if that would be needed. _

Desktop default codepage is returned by Windows `GetACP()` function. It would be 1251 for Russian, 1250 and 1252 for East and West Europe, and so forth. Potentially can be 65001 for UTF-8, but i did not see it used...

_Technically, every Win32 process and even thread can have its own non-default codepage, but i do not think i ever saw it._

Interesting that as modern desktop Linux had settled on UTF-8, earlier UNIXes chose `koi8-r` or `windows-1251` for Russian GUI. It is not customary now, but probably still possible. Should be signalled by the "nested" number of environment variables, LC_xxx or LOCALE.xxx - i do not remember anymore, but again, this can be found if needed. 

FPC 2.x as far as i heard had a mess about `string` variables charsets, their RTL and other libraries were ignorant of charsets, yet there was no standard. FPC 3.x fixed use of UTF-8

Just like 2009+ Delphi fixed upon UTF-16 `string == UnicodeString` type.

Now, back to the screenshot.
It is a typical easily recognizable case when UTF-8 text bytestream is passed for windows-1251 bytestream.
Which signals two things:

 * GCE-generated application uses non-Unicode "ANSI" (more corrcectly, MBCS) subset of Win32 API (WhateverFunctionNameA ones)
 * Castle Engine does not transcode FPC charset (UTF07) to Windows MBCS charset, at least for Main Form caption
 * ...but probably for other UI elements too
 * chances are, it would be the same with non-UTF-8 Linux setups...

P.S. yeas, it also affects buttons, so probably all the UI widgets.

![2022-01-09 16_08_39- _gamestatemenu castle-user-interface  WaterSonar _ Castle Game Engine](https://user-images.githubusercontent.com/937774/148683513-47a3a58c-9381-40d7-98e9-75e33acf74c3.png)

![2022-01-09 16_11_15-Р—РІСѓРє РІ РІРѕРґРµ](https://user-images.githubusercontent.com/937774/148683525-bedde891-72ec-4284-b19e-06653cf1434c.png)

**Comment by the-Arioch** [comment]
I kinda am buffled...

The project caption is BOTH contained in the `CastleEngineManifest.xml` nd in the `castleautogenerated.pas`

What about DRY ?  What is the difference between that value being in one file and in another ?

**Comment by the-Arioch** [comment]
In Delphi the evaluate/modify window can coerce results, so it is trivial to make a short HX dump.

There is a RadioGroup with format modifiers, or it can be added as flags to the expression itself, https://docwiki.embarcadero.com/RADStudio/Sydney/en/Evaluate/Modify

There is no that radiogroup in Laz, and when io try to enter `PCardinal(@FApplicationProperties.Caption[1])^,x` - it complains about failure to parse comma...

Is there not a way to do it as simpel as in Delphi?

**Comment by michaliskambi** [comment]
> But then the Castle sets main window Caption to that byte stream as is, without converting.

Thanks for the explanation.

You don't need to explain how encodings works, we know how they work in FPC and Delphi :) Your explanation wasn't also really correct -- FPC has code-page aware AnsiString. In CGE we configure it such that we actually just assume that AnsiString contains just always UTF-8, this is exactly what Lazarus LCL does too. See https://castle-engine.io/coding_conventions#_most_code_should_use_just_string_and_be_prepared_that_it_is_8_bit_on_fpc_and_16_bit_on_delphi_if_writing_to_stream_use_8_bit_ansistring .

Your error just means that we didn't convert caption from UTF-8 -> Windows codepage correctly. Indeed, our `castlewindow_winapi.inc` just does passes `PChar(GetWholeCaption)`, we should add there conversion UTF-8 -> Ansi. I'll do it shortly.

>  yeas, it also affects buttons, so probably all the UI widgets.

Your button is an unrelated issue. By default, we don't load non-ASCII characters from fonts. Extend data/CastleSettings.xml (see https://castle-engine.io/manual_castle_settings.php ) to load non-ASCII characters in the default font (see property `sample_text` there).

You could also create other fonts, and assign there `LoadCharacters` to load non-ASCII characters. See https://castle-engine.io/wp/2021/07/11/font-classes-upgraded-and-available-in-cge-editor/ (the docs on https://castle-engine.io/manual_text.php should be updated with this information too, right now they are a little outdated).

> The project caption is BOTH contained in the CastleEngineManifest.xml nd in the castleautogenerated.pas

This is again an unrelated question. 

Please, don't overload the issues / forum threads with multiple questions. It make it really hard to follow and answer :) The answer to this "indeed, ideally we should not need to duplicate CastleEngineManifest.xml information in a castleautogenerated.pas, or hide castleautogenerated.pas more from you". We don't do it yet, to support easy complications from outside of CGE editor (you can use Lazarus or Delphi without CGE build tool / editor too).

I hope to improve this some day, but it depends on other factors.

> Is there not a way to do it as simpel as in Delphi?

That's an unrelated Lazarus debugger issue.

**Comment by the-Arioch** [comment]
Now i feel really lost... Are strings in Lazarus 0-based or 1-based???
I just can wrap my mind over these 4 screenshots:
![2022-01-09 16_36_54-Вычислить_Изменить](https://user-images.githubusercontent.com/937774/148684537-7f56d86a-ceb5-40f6-8eaa-14cbf0c90453.png)
![2022-01-09 16_36_34-Вычислить_Изменить](https://user-images.githubusercontent.com/937774/148684538-d849d307-491c-4a0b-97cb-45a9c22ae896.png)
![2022-01-09 16_36_08-Вычислить_Изменить](https://user-images.githubusercontent.com/937774/148684540-e1798dce-e92b-4ba3-92e9-72bb9a0f4ab8.png)
![2022-01-09 16_35_50-Вычислить_Изменить](https://user-images.githubusercontent.com/937774/148684541-dbd010d1-aa68-4ef4-81af-ebea26a04ba3.png)

**Comment by the-Arioch** [comment]
> You don't need to explain how encodings works, we know how they work in FPC and Delphi :)

That's untypical for westerners :-) 

> I'll do it shortly.

Okay, leaving it to you then, i tried to find the place myself, havign the specific data that gets garbled. But if you know where to patch, that is one less burden for me (and one more for you :-) )

**Comment by michaliskambi** [comment]
> Now i feel really lost... Are strings in Lazarus 0-based or 1-based???

Strings in FPC and Lazarus are 1-based, as in FPC.

I suppose what you show is an error in expressions you input in Lazarus "watches". Lazarus uses GDB under the hood and it doesn't really understand Pascal expressions. Lazarus team has fixes/workarounds to hide this, from what I know, but indeed they don't work 100%. So sometimes internal things "creep in", like the fact that for GDB, strings are like C strings, i.e. 0-based array of 8-bit chars.

You can submit such issues to Lazarus.

**Comment by the-Arioch** [comment]
> Indeed, our castlewindow_winapi.inc just does passes PChar(GetWholeCaption), we should add there conversion UTF-8 -> Ansi.

1. Thanks for pointing at the specific file
2. It probaly is not Windows-specific. Checking non-UTF-8 Linux/X configurations is probably needed too
3. "conversion UTF-8 -> Ansi" seems wrong approach

I don't nkow if compatibility with Microsoft Windows 98 is still any priority, but even if it is - there is Microsoft Layer for Unicode for it. 

Modern Windows use UTF-16 as their base charset. So you actually here talk about 3 operations instead of 1.

Option 1:

```
var CaptionSource: UTF8String;
    CaptionTmp: WideString; // or UnicodeString
    CaptionPtr: PWideChar;
....
   CaptionTmp := CaptionSource;  // conversion 1
   CaptionPtr := @CaptionTmp[1]; 
   Windows.CreateWindowW( ... CaptionPtr ...); // using Windows API directly
```

Option 2:

```
var CaptionSource: UTF8String;
    CaptionTmp: array of WideChar;
    CaptionAnsi: AnsiString; // or UnicodeString
    CaptionPtr: PAnsiChar;
    BufLen: cardinal;
....
   BufLen := 0;
   Windows.MultiByteToWideChar(CP_UTF8, @CaptionSource[1], nil, BufLen);
   SetLength(CaptionTmp, BufLen); // 2*Length(CaptionSource) would work 99%, but not 100%
   Win32Check( Windows.MultiByteToWideChar(CP_UTF8, @CaptionSource[1], @CaptionTmp[0], BufLen) ); // conversion 1
   SetCodePage( CaptionAnsi, GetACP() );
   BufLen := Length(CaptionSource);
   SetLength( CaptionAnsi, BufLen );
   Win32Check( Windows.WideCharToMultiByte(@CaptionTmp[0], CP_ANSI, @CaptionAnsi[1], BufLen) ); // conversion 2
   SetLength( CaptionAnsi, BufLen );
   CaptionPtr := @CaptionAnsi[1]; 
   Windows.CreateWindowA( ... CaptionPtr ...); // using Windows API wrapper

....somewhere inside Microsoft GDI.DLL

   MultiByteToWideChar(CP_ANSI, CaptionPtr, ..... ) );  // conversion 3
   return CreateWindowW( .... );
```

**Comment by the-Arioch** [comment]
> Indeed, our castlewindow_winapi.inc just does passes PChar(GetWholeCaption), we should add there conversion UTF-8 -> Ansi.

1. Thanks for pointing at the specific file
2. It probaly is not Windows-specific. Checking non-UTF-8 Linux/X configurations is probably needed too
3. "conversion UTF-8 -> Ansi" seems wrong approach

I don't nkow if compatibility with Microsoft Windows 98 is still any priority, but even if it is - there is Microsoft Layer for Unicode for it. 

Modern Windows use UTF-16 as their base charset. So you actually here talk about 3 operations instead of 1.

Option 1:

```
var CaptionSource: UTF8String;
    CaptionTmp: WideString; // or UnicodeString
    CaptionPtr: PWideChar;
....
   CaptionTmp := CaptionSource;  // conversion 1
   CaptionPtr := @CaptionTmp[1]; 
   Windows.CreateWindowW( ... CaptionPtr ...); // using Windows API directly
```

Option 2:

```
var CaptionSource: UTF8String;
    CaptionTmp: array of WideChar;
    CaptionAnsi: AnsiString; // or UnicodeString
    CaptionPtr: PAnsiChar;
    BufLen: cardinal;
....
   BufLen := 0;
   Windows.MultiByteToWideChar(CP_UTF8, @CaptionSource[1], nil, BufLen);
   SetLength(CaptionTmp, BufLen); // 2*Length(CaptionSource) would work 99%, but not 100%
   Win32Check( Windows.MultiByteToWideChar(CP_UTF8, @CaptionSource[1], @CaptionTmp[0], BufLen) ); // conversion 1
   SetCodePage( CaptionAnsi, GetACP() );
   BufLen := Length(CaptionSource);
   SetLength( CaptionAnsi, BufLen );
   Win32Check( Windows.WideCharToMultiByte(@CaptionTmp[0], CP_ANSI, @CaptionAnsi[1], BufLen) ); // conversion 2
   SetLength( CaptionAnsi, BufLen );
   CaptionPtr := @CaptionAnsi[1]; 
   Windows.CreateWindowA( ... CaptionPtr ...); // using Windows API wrapper

....somewhere inside Microsoft GDI.DLL

   MultiByteToWideChar(CP_ANSI, CaptionPtr, ..... ) );  // conversion 3
   return CreateWindowW( .... );
```

**Comment by the-Arioch** [comment]
Damn... it was crazy... To think that for low-level debugging that commercial Delphi is better than FLOSS Lazarus.

Also, whoever thought `SetMultiByteConversionCodePage(CP_UTF8)` was good idea...
...let's say was overly optimistic. One instantly broke text conversions by assignubg ( `CaptionTmp := CaptionSource;` );

Well, okay, the title seems to work now FPC/Win64, both with Unicode and ANSI API.

To tired to try FPC/Win32, Delphi/Win32 and Delphi/Win64.
But at least your primary Windows target should work now.

See 
[0001-Implementing-non-Latin-letters-for-Window-Caption.patch.zip](https://github.com/castle-engine/castle-engine/files/7836063/0001-Implementing-non-Latin-letters-for-Window-Caption.patch.zip)

P.S. that was what it used to look like. I almost run crazy...

![2022-01-10 00_18_12-UTF16 Encode Decode - Convert String to UTF16 - ConvertCodes - Vivaldi](https://user-images.githubusercontent.com/937774/148702772-4d01e015-69df-43f8-8f2e-74ce0c31bfdd.png)
![2022-01-10 00_11_28-Window Detective](https://user-images.githubusercontent.com/937774/148702774-98c69ff3-6084-4c2b-9906-f1627f73dd9d.png)
![2022-01-10 00_19_19-Lazarus IDE v2 2 0 - WaterSonar (идёт отладка  )](https://user-images.githubusercontent.com/937774/148702771-3c422cef-648d-4a2d-93d5-37897b6d50cf.png)

**Comment by michaliskambi** [comment]
Ah, I just did the same thing... 

I'm guessing that, same as you, I though _"this should be trivial, esp. as we have that CASTLE_WINAPI_UNICODE as a starting point"_.  Then I'm guessing that, same as you, I spent absurd amount of time trying to figure out why `CreateWindowExW` still expects caption encoded as 8-bit (!!!) not 16-bit :)

The solution I found was DefWindowProc -> DefWindowProcW. I see you used `CallWindowProcW` in an equivalent place.

Well, WinAPI docs warn to not mix ANS and Unicode API. Though I didn't except them to be so weird, and autodetect whether you use ANSI DefWindowProc or not -> to treat the caption parameter to `CallWindowProcW` differently.

I'm going with my version, as you can see I completely removed non-CASTLE_WINAPI_UNICODE version, and I upgraded a bunch of other calls (related to cursors, menu) to Unicode versions too. I checked with FPC/Win64 and Delphi/Win64.

I also added BOM to generated castleautogenerated, otherwise at least Delphi was trying to interpret the caption (generated there in UTF-8) as ANSI, and everything further was bad.

Fixed by

- https://github.com/castle-engine/castle-engine/commit/c8db5c2383604c4c16b187f06e694a5204b4b1ff
- https://github.com/castle-engine/castle-engine/commit/baf240b6a270d6cf09b3dd4549d465f7d286707d

**Comment by michaliskambi** [comment]
This is the minimal WinAPI test program I used to find the issue: https://gist.github.com/michaliskambi/870e289a0e193df60343aa40783c6769 . It creates WinAPI window, without any CGE dependencies (so that I knew the weirdness doesn't have anything to do with anything else in CGE).

**Comment by the-Arioch** [comment]
Frankly, i think you was too optimistic, as we instantly see in #364 

I still think that

* you better use `CallWindowProc` as MSDN seems to require. Maybe it is Win16 remain, maybe it only matters in niche cases, but before you hit it you do not know. At very least, leave behind the comment with reference to MSDN and function names - year in future no one would remember there even was a doubt.
* non-Unicode version still has value as a fall-back, for debugging purposes for example. For example, to run ANSI and Unicode versions side by side comparing behavior differences.
* I agree ifdef's everywhere is not elegant. (Though i think FPC habit to use includes everywhere is not much better, and so are Castle's ifdef-ladden includes too. PAscal is not Java, for good or for bad). 

So, i think if `CASTLE_WINAPI_UNICODE` is an enemy, then we better:

* offset it to some Windows-global include file. So the ANSI/Unicode mode would be switched on/off in single place globally.
* to get rid of many small ifdefs in `castlewindow_winapi.inc` - make one big ifdef with different declarations of data types (like string and PChar in Delphi, like TCHAR and PTCHAR in MS VC++) and inline/forward functions.

I wonder if older Delphi versions (pre-Unicode and pre-advanced records) are to be targeted.

Something like this
```
type
{$IfDEf CASTLE_WINAPI_UNICODE}
  TWinChar = WideChar;
  TWinString = WideString;
// UnicodeString could work too, but we do not actually need ARC and COW here.
{$Else} 
  TWinChar = AnsiChar;
  TWinString = AnsiString;
{$EndIf}
  PWinChar = ^TWinChar;

  RWinString = record // make the compiler manage buffer runtime for C-based GDI API
    Buffer: TWinString;
    class operator implicit(const Src: RWinString): TWinString; inline;
    class operator implicit(const Src: RWinString): PWinChar; inline;

    // convert back, dunno if FPC/Delphi would allow such an overloading...
    // class operator implicit(const Src: RWinString): TUTF8String; inline; 
  end;

 class operator implicit(const Src: RWinString): PWinChar; 
 begin
   Result := @Self.Buffer[1];
 end;

 class operator implicit(const Src: RWinString): TWinString; 
 begin
   Result := Self.Buffer;
 end;

 function TextCastleToWindows(const Src: string): RWinString;
 begin
{$IfDEf CASTLE_WINAPI_UNICODE}
    Result.Buffer := Src; // make the compiler auto-convert it
{$Else} 
    // here we probably should tell Unicode Delphi from FPC and pre-Unicode Delphi
    Result.Buffer := Src;
    SetCodePage(Result.Buffer, GetACP(), True);  // only needed for UTF-8 FPC
{$EndIf}
 end;

....

function {local} CreateWindowEx(.....): THandle; inline;  // AFAIR HWND is obsolete in Win32
begin
{$IfDEf CASTLE_WINAPI_UNICODE}
    Result := Windows.CreateWindowExW( ... );
{$Else} 
    Result := Windows.CreateWindowExA( ... );
{$EndIf}
end;

.... later in main code ...

 H_wnd := CreateWindowEx(dwExStyle,
    TextCastleToWindows(Application.WndClassName), 
    TextCastleToWindows(GetWholeCaption),
   ....
```

**Comment by the-Arioch** [comment]
> DefWindowProcW. I see you used CallWindowProcW in an equivalent place

to clarify, i used BOTH

and i commented in the patch that in my 5 minutes test i did not see difference with or withour `CallWindowProc` it was unicodeness that mattered.

however i prefer here to be uber-cautious and follow MSDN warning and follow the contract by wrapping Window Function in `CallWindowProc` proxy. Maybe it would never matter, but better safe than sorry.

**Comment by michaliskambi** [comment]
> you better use CallWindowProc as MSDN seems to require

I looked into that.

I do not interpret https://docs.microsoft.com/en-us/windows/win32/api/winuser/ns-winuser-wndclassa as telling me to execute CallWindowProcW instead of DefWindowProcW directly. The way I understand it, it tells me to not execute my own callback (`WindowProc` in `castlewindow_winapi.inc`) directly, but rather always use `CallWindowProc` -- OK.

Reading https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-callwindowproca ("Use the CallWindowProc function for window subclassing."), https://docs.microsoft.com/en-us/previous-versions/windows/desktop/legacy/ms633573(v=vs.85) (does say whether to use DefWindowProc or CallWindowProc for fallback), https://docs.microsoft.com/pl-pl/windows/win32/api/winuser/nf-winuser-defwindowproca?redirectedfrom=MSDN ("Calls the default window procedure to provide default processing for any window messages that an application does not process. ") -> this leads me to believe that calling DefWindowProc is OK.

Looking at what LCL (win32 widgetset) is doing: CallDefaultWindowProc.. inconclusive, they use `DefWindowProc` to get default behavior (which looks similar to our case -- when the message is unhandled call `DefWindowProc`) though they also have code to call `CallWindowProcW` in some cases.

I will keep this in mind for the future, but I think that our usage of DefWindowProc is now correct. So I'll leave it as it is.

> non-Unicode version still has value as a fall-back, for debugging purposes for example. For example, to run ANSI and Unicode versions side by side comparing behavior differences.

The ANSI version of castlewindow_winapi.inc would lie mostly untested, in my opinion. And it would be a burden -- every modification to CastleWindow WinAPI backend would need to take it into account, to use only-Unicode or only-ANSI WinAPI routines consistently.

So, no -> the maintenance burden is too big, to maintain an ANSI version of WinAPI usage when we explicitly know we don't want to use it, and Unicode version of WinAPI is the future. The gain (to occasionally compare ANSI with Unicode version) is too low.

Whether we do this by ifdefs local to castlewindow_winapi.inc, or we do it by creating another file with a compatibility layer -> in both cases, we would introduce lots of additional maintenance in the future, lots of changes would need to take care of both ANSI and Unicode variant.

To add more, I see we already have been using `PeekMessageW` (Unicode) even before last modifications (upgrade to CreateWindowExW), i.e. we were using `PeekMessageW` mixed with ANSI methods, in `Handle_WM_KEYDOWN`. We needed that to read input as Unicode easily. So our ANSI code was never really 100% correct... and I definitely don't want now to add work to fix the ANSI approach.

Going to Unicode version of WinAPI was a good decision, we should just stick to it.

**Comment by the-Arioch** [comment]
Well, this use of `UTF8Decode` everywhere is still bad...

Delphi XE2 system.pas:
```
{ WideString <-> UTF8 conversion }

function UTF8Encode(const WS: WideString): RawByteString; overload;
function UTF8Encode(const US: UnicodeString): RawByteString; overload;
function UTF8Encode(const A: RawByteString): RawByteString; overload;
function UTF8EncodeToShortString(const WS: WideString): ShortString; overload;
function UTF8EncodeToShortString(const US: UnicodeString): ShortString; overload;
function UTF8EncodeToShortString(const A: RawByteString): ShortString; overload;


function UTF8Decode(const S: RawByteString): WideString; deprecated 'Use UTF8ToWideString or UTF8ToString';


function UTF8ToWideString(const S: RawByteString): WideString; inline;
function UTF8ToUnicodeString(const S: RawByteString): UnicodeString; overload;
function UTF8ToUnicodeString(const S: ShortString): UnicodeString; overload;
function UTF8ToUnicodeString(const S: PAnsiChar): UnicodeString; overload;
function UTF8ToString(const S: RawByteString): string; inline; overload;
function UTF8ToString(const S: ShortString): string; inline; overload;
function UTF8ToString(const S: PAnsiChar): string; inline; overload;
function UTF8ToString(const S: array of AnsiChar): string; overload;
//function UTF8ToUnicodeString(const S: ShortString): UnicodeString; overload;

{ Ansi <-> UTF8 conversion }

function AnsiToUtf8(const S: string): RawByteString;
function Utf8ToAnsi(const S: RawByteString): string;

{ OLE string support procedures and functions }

function OleStrToString(Source: PWideChar): UnicodeString;
procedure OleStrToStrVar(Source: PWideChar; var Dest: AnsiString); overload;
procedure OleStrToStrVar(Source: PWideChar; var Dest: UnicodeString); overload;
function StringToOleStr(const Source: AnsiString): PWideChar; overload;
function StringToOleStr(const Source: UnicodeString): PWideChar; overload;
```

Those functions are remnants of Delphi-2007 and prior ones, where there was not codepages in strings.
They were left for source compatibility, but their behaviour is... non intuitive.

In Delphi it is much cleaner and more predictable to just use explicit typecasts and `SetCodepage`;
However, LCL-like setting "zero" codepage to UTF-8 breaks this logic too.

`PWideChar(UTF8Decode(GetWholeCaption)),` - but `GetWholeCaption` is *already* UTF-16 !!!

So, for Unicode Delphi ( and it seems you do not target pre-Unicode ones? ) it would most probably lead to non-Latin1 text deletion.

```
var s: string; // == UnicodeString; or WideString == BSTR

s := 'DEFДЕЖ';
s := UTF8Decode(s);

```

What would happen there?
1. `s` would be coerced (with a warning) into AnsiString, using 1251 codepage.
2. then cp1251 text would be misinterpreted as UTF-8 text
3. Latin1 letters have the same binary representation in 1251 and UTF-8, but not cyrillics
4. thanks to "interesting" design of `TEncoding` when meeting decoding error just silently returns empty string. No error, no partial conversion - just empty string;

**Comment by michaliskambi** [comment]
> Well, this use of UTF8Decode everywhere is still bad...

Just 3 calls to UTF8Decode, all when setting caption. Don't exaggerate by "use of everywhere...".

> So, for Unicode Delphi ( and it seems you do not target pre-Unicode ones? ) it would most probably lead to non-Latin1 text deletion.

Did a simple test with Delphi 11, and it is not true. The Russian characters in S are preserved OK, I can then display the resulting string (e.g. on window caption) and it is OK.

> PWideChar(UTF8Decode(GetWholeCaption)), - but GetWholeCaption is already UTF-16 !!!

In case of Delphi, the `UTF8Decode` call indeed should do just nothing. Luckily, my above test confirms that it does nothing. 

And we need the `UTF8Decode` on FPC.

So these 3 calls to `UTF8Decode` work reliably, and give us correct results with both FPC and Delphi. So I'll keep them.

**Comment by the-Arioch** [comment]
there are a lot of calls in JSON code, maybe in others repos too

What is minimum Delphi to be supported?

Does Delphi 11 register do-nothing overloads for UTF8Decode, or does it
convert to AnsiString and back?

XE2 i quoted, there is just no UTF8Decode(UnicodeString) overload declared

**Comment by michaliskambi** [comment]
> there are a lot of calls in JSON code, maybe in others repos too

If you speak abuot CGE/FPC JSON code -- there are 2 UTF8Decode calls, in fpjson.pas (this is FPC unit, though we use it with Delphi too).

There's a number of UTF8Decode calls in CastleXmlUtils, because FPC API uses WideStrings, inconsistently with rest of RTL.

I do not consider these calls a problem -- I never experienced them being a bottleneck in anything. I do not know what exactly Delphi does with them, but experiments above show that results are as expected, i.e. as if UTF8Decode did nothing in this case. So there's no need to patch it for Delphi, it seems.

> What is minimum Delphi to be supported?

https://castle-engine.io/supported_compilers.php#section_delphi

**Comment by the-Arioch** [comment]
![2022-01-14 09_35_03-Project46 - Delphi XE2 - Unit27  Running](https://user-images.githubusercontent.com/937774/149462677-327e8b6b-b3ef-4f4c-a1fb-9bf9217bf219.png)
![2022-01-14 09_34_33-Project46 - Delphi XE2 - Unit27  Running](https://user-images.githubusercontent.com/937774/149462688-e9d22d93-1c0b-49d2-a83d-507d3dbc590e.png)

**Comment by the-Arioch** [comment]
I understand about dependencies, but think about mORMot as JSON implementation, they are speed maniacs, so it might fit well with game engine. If you bothered to copy the whole vampyre, then extract one json unit from mORMot might be easier. And they were very strong Delphi players that rcecently moved to FPC, so it would have greate support for both platforms.

**Comment by the-Arioch** [comment]
Also consider making `Application.WndClassname` into `AnsiString` due to #364 

being technical identifier you do not need anything but Latin in it, this type change would not change a thing for FPC, but would avoide double-conversion on Delphi (Unicode Delphi but you do not support any other)

Also, i would remove Delphi 2009 from that page, in my experience the generics there were absolutely unsuable, and the documentation was not mere wrong but absurd. I was _very_ disappointed then had to work with D2009 for about a month.

**Comment by michaliskambi** [comment]
As for your test: do it within the CGE application. We do `SetMultiByteConversionCodePage(CP_UTF8);` in CastleUtils initialization.

I tested it, pasted your example earlier

```
var s: string; // == UnicodeString; or WideString == BSTR

s := 'DEFДЕЖ';
s := UTF8Decode(s);
```

into a CGE example, set `Window.Caption := S` and it works (Russian chars are rendered OK). Yes, it likely converts to UTF-8 and then back to UTF-16 needlessly on Delphi.

> I understand about dependencies, but think about mORMot as JSON implementation

We use FPC JSON implementation, including FpJsonRtti, also with Delphi. I do not plan to change it now, as I do not see a reason.

> Also consider making Application.WndClassname into AnsiString due to #364
>
> being technical identifier you do not need anything but Latin in it, this type change would not change a thing for FPC, but would avoide double-conversion on Delphi (Unicode Delphi but you do not support any other)

No. We'll keep WndClassname for consistency with other strings we pass to WinAPI there. 

And there's no double conversion related to WndClassname. Since it's UTF-16 `UnicodeString` in both FPC and Delphi, we just pass `PWideChar(WndClassName)` to API calls. So there's actually 0 conversions related to WndClassname in current approach.

**Comment by the-Arioch** [comment]
> We do SetMultiByteConversionCodePage(CP_UTF8);

Would check tonight, thought it was FPC-specific trick...
No i really warry how this would co-exist with VCL... Not personally, i do not actually plan to use Delphi with GC, but still.

> Since it's UTF-16 UnicodeString in both FPC and Delphi

That's ok too, just skipping `UTF8Decode` completely then.

Perhaps i (wrongly) inferred the type from existense of #364 and from `GetWholeCaption` function, or remember form earlier commits. Would check tonight, too.

**Comment by the-Arioch** [comment]
> Yes, UTF8Decode likely converts to UTF-8 and then back to UTF-16 needlessly on Delphi -- not a problem right now, in my opinion, it actually is done extremely rarely in actual applications.

Just add 5-line no-brainer into Windows .inc 

```
{$if SizeOf(char) = 2}
function UTF8Decode(const src: UnicodeString): UnicodeString; overload; inline;
begin
  Result := src;
end;
{$ifend}
```

Is `CreateWindow[Ex]` the only Win32 API function used directly, which has text params thus needing cast to `PWideChar`? If number of them are (would be) used i'd still vote for the adv-record wrapper above. But for a single API call that would be overengineering indeed.

Is UTF-16 only used on Windows with every other platform API using UTF-8? 
_i still remember Linux days before UTF-8 and that still is technically possible_

**Comment by the-Arioch** [comment]
> We do SetMultiByteConversionCodePage(CP_UTF8);

Confirmed. Works in XE2 at least for basic stuff

**Comment by michaliskambi** [comment]
> function UTF8Decode(const src: UnicodeString): UnicodeString; overload; inline;

Overloading UTF8Decode like this means you risk not actually applying this to all the calls. And you would actually miss `CreateWindowExW` in `castlewindow_wgl.inc` if you just pasted it in implementation section of `castlewindow_winapi.inc`. If we decide that this optimization is worth pursuing, then it should be done consistently in all WinAPI includes.

I instead added new `StringToUtf16`, which acts as a wrapper over UTF8Decode (for FPC) and nothing (for Delphi) in https://github.com/castle-engine/castle-engine/commit/550a13278234ba16ce94fce3e88ca4c767e1209d , and used in WinAPI code.

> Is UTF-16 only used on Windows with every other platform API using UTF-8?
> i still remember Linux days before UTF-8 and that still is technically possible

Everything else except WinAPI is using UTF-8.

**Comment by the-Arioch** [comment]
> I instead added new StringToUtf16, which acts as a wrapper
Great. That is a better approach, self-documenting and safe.
